### ISSUE TITLE
Overlapping definition names creates ambiguous matches

### ISSUE THREAD

**Issue by rtkjweeks** [daschl/grok#22]
I'm in the process of porting a python system into rust, and am using your grok library, but am finding some differences between the implementation.

The following rust code shows is an example of what I'm seeing:

```
use grok::Grok;

fn main() {
    let mut grok = Grok::default();

    let lines: [&str; 18] = [
        r"MSG=\S.*",
        r"DAY=\d{1,2}",
        r"MONTH_NUMBER=\d{1,2}",
        r"MONTH_SHORT_NAME=[A-Za-z]{3}",
        r"MONTH=MONTH_NUMBER|MONTH_SHORT_NAME",
        r"YEAR=\d{4}",
        r"HOUR=\d{2}",
        r"MINUTE=\d{2}",
        r"SECOND=\d{2}",
        r"MICROSECOND=\d{1,6}",
        r"DATESTAMP_1=%{YEAR:year}-%{MONTH_NUMBER:month}-%{DAY:day}",
        r"DATESTAMP_3=%{MONTH_SHORT_NAME:month}[ ]* %{DAY:day}( %{YEAR:year})?",
        r"DATESTAMP_3_NO_YEAR=%{MONTH_SHORT_NAME:month}[ ]* %{DAY:day}",
        r"TIMESTAMP=%{HOUR:hour}:%{MINUTE:minute}:%{SECOND:second}([.,]%{MICROSECOND:microsecond})?",
        r"DATE_AND_TIME_1=%{DATESTAMP_1:datestamp}T%{TIMESTAMP:timestamp}",
        r"DATE_AND_TIME_3=%{DATESTAMP_3:datestamp} %{TIMESTAMP:timestamp}",
        r"DATE_AND_TIME_5=%{DATESTAMP_3_NO_YEAR} %{TIMESTAMP:timestamp} %{YEAR:year}",
        r"COMBINED_DATE_AND_TIME=(%{DATE_AND_TIME_1}|%{DATE_AND_TIME_5}|%{DATE_AND_TIME_3})",
    ];


    for grok_pattern in lines {
        if let Some(first_space) = grok_pattern.find('=') {
            let (name, value) = grok_pattern.split_at(first_space);
            grok.insert_definition(name, &value[1..]);
        }
    }

    match grok.compile(r"%{COMBINED_DATE_AND_TIME:date_and_time}[ ]{1,}%{MSG:msg}", true) {
        Ok(pattern) => {
            let log = "2017-01-12T11:57:02.206 Testing";
            if let Some(matches) = pattern.match_against(&log) {
                println!("Found match");
                for (name, value) in matches.iter() {
                    println!("  {} -> {}", name, value);
                }
            }
        },
        Err(e) => println!("{}", e)
    }
}
```

This will product output of:

```
Found match
  hour ->
  name15 ->
  day ->
  name3 -> 2017
  minute ->
  name6 -> 11:57:02.206
  name16 ->
  name20 ->
  datestamp ->
  year ->
  name8 -> 57
  date_and_time -> 2017-01-12T11:57:02.206
  microsecond ->
  name5 -> 12
  second ->
  name9 -> 02
  timestamp ->
  msg -> Testing
  name13 ->
  name14 ->
  month ->
  name10 -> 206
  name7 -> 11
  name18 ->
  name2 -> 2017-01-12
  name19 ->
  name4 -> 01
  name17 ->
```

So the pattern matched, but I'm unable to extract the components of the date, such as the year, because `matches.get("year")` will return None.  The year is actually contained in `matches.get("name3")` (presumably because the patterns `DATE_AND_TIME_1`, `DATE_AND_TIME_3` and `DATE_AND_TIME_5` each contain a `year`... but only one of them should match, so I'm expecting only one value of `year`.

Is there a way to lookup key `year` to `name3`?
Or, preferably, have `matches.get("year")` return the matched year?

**Comment by rtkjweeks** [comment]
I see now that I misinterpreted part of the implementation, and it looks like, for my use case, I probably can add a look-up map in the reverse direction of Pattern::names.

I'm curious, though, why Matches::iter iterates over the aliases map; shouldn't it iterate over only the matched names?  In my case, it results in iterating over a bunch of names with no values.

**Comment by daschl** [comment]
I think you are correct that the matches iterator, well, should only iterate over the actual matches. I think we can make the change but it would have to be a new major version (or we introduce some kind of switch for now) since it would be a backwards incompatible change.

**Comment by daschl** [comment]
Fixed this in main, will be part of the next 2.0 release since it is a breaking change but makes a lot of sense.

**Comment by daschl** [comment]
2.0.0 is released, want to try it out?
